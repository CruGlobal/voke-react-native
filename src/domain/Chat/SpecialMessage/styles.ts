import { StyleSheet } from 'react-native';
import theme from 'utils/theme';

const sharedStyles = {
  avatar: {
    position: 'absolute',
    bottom: 0,
    height: 25,
    width: 25,
    borderRadius: 50,
  },
};

const styles = StyleSheet.create({
  mainQuestionCard: {
    // paddingTop: theme.spacing.l,
    paddingHorizontal: theme.spacing.l,
    paddingBottom: theme.spacing.l,
  },
  mainQuestionContainer: {
    paddingHorizontal: 20,
  },
  mainQuestion: {
    width: '100%',
    paddingVertical: theme.spacing.l,
    paddingHorizontal: theme.spacing.m,
    borderTopRightRadius: 10,
    borderTopLeftRadius: 10,
    backgroundColor: theme.colors.accent,
  },
  questionText: {
    textAlign: 'center',
    color: theme.colors.white,
    fontSize: theme.fontSizes.l,
    lineHeight: theme.fontSizes.l * 1.3,
  },
});

export default styles;
