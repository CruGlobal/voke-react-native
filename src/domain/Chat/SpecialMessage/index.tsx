import InteractiveElement from 'domain/Chat/InteractiveElement';

import React, { useState } from 'react';
import { View, Text } from 'react-native';
import { TAdventureSingle, TMessage, TStep } from 'utils/types';
import Flex from 'components/Flex';

import styles from './styles';

interface Props {
  message: TMessage;
  nextMessage: TMessage | null;
  kind: TStep['kind'];
  step: TStep;
  onFocus: (answerPosY: number) => void;
  adventure: TAdventureSingle;
}

const SpecialMessage = (props: Props): React.ReactElement => {
  const { adventure, message, nextMessage, kind, step, onFocus } = props;
  const [answerPosY, setAnswerPosY] = useState(0);

  let selectedAnswer =
    (((message?.metadata || {}).answers || []).find((i) => i?.selected) || {})
      .key || '';

  // If current message is a question box and next message is answer,
  // render the next message here (https://d.pr/i/YHrv4N).
  if (
    kind === 'question' &&
    !selectedAnswer &&
    message?.metadata?.vokebot_action === 'journey_step' &&
    nextMessage?.content &&
    nextMessage?.direct_message
  ) {
    selectedAnswer = nextMessage?.content;
  }

  return (
    <View
      style={styles.mainQuestionCard}
      onLayout={({ nativeEvent }): void => {
        // Calculate vertical offset to be used on the answer field focus.
        const layout = nativeEvent?.layout;
        if (layout && layout?.y && layout?.height) {
          setAnswerPosY(layout.y);
        }
      }}
    >
      <Flex direction="column" style={styles.mainQuestionContainer}>
        {kind === 'multi' || kind === 'question' ? (
          /* Message question area: */
          <Flex
            direction="column"
            style={styles.mainQuestion}
            align="center"
            justify="center"
          >
            <Text style={styles.questionText}>
              {(message?.metadata || {}).question || null}
            </Text>
          </Flex>
        ) : null}
        {/* Message input field / options / actions: */}
        <InteractiveElement
          kind={kind}
          adventure={adventure}
          step={step}
          internalMessage={message}
          defaultValue={selectedAnswer}
          onFocus={(): void => {
            onFocus(answerPosY);
          }}
        />
      </Flex>
    </View>
  );
};

export default SpecialMessage;
