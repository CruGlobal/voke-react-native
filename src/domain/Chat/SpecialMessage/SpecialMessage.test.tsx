import React from 'react';
import { fireEvent } from '@testing-library/react-native';
import '@testing-library/jest-native';
import {
  mockUser,
  mockStepDecisionInactive,
  mockMessageBinaryDefault,
  mockMessageBinarySelected,
  groupAdventureMock,
  mockMessageQuestion,
} from 'mocks/vokeDataMocks';
import renderWithContext from 'utils/testUtils';

import SpecialMessage from './index';

const initialState = {
  auth: { user: mockUser },
  data: {
    adventureStepMessages: {
      [mockStepDecisionInactive.id]: {},
    },
  },
};

beforeEach(() => {
  // https://github.com/facebook/jest/issues/6434#issuecomment-525576660
  jest.useFakeTimers();
});

it('renders correctly - kind "question" - not answered', () => {
  const { toJSON, queryByText, queryByTestId } = renderWithContext(
    <SpecialMessage
      message={mockMessageQuestion}
      nextMessage={null}
      kind="question"
      adventure={groupAdventureMock}
      step={mockStepDecisionInactive}
      onFocus={(): void => {
        // void
      }}
    />,
    { initialState },
  );

  expect(toJSON()).toMatchSnapshot();

  expect(queryByText(/What led you to this response today/i)).toBeTruthy();
  expect(queryByTestId('inputEnterAnswer')).toBeTruthy();
  expect(queryByTestId('ctaSendAnswer')).toBeTruthy();
});

it('renders correctly - kind "question" - answered', () => {
  const { toJSON, queryByText, queryByTestId } = renderWithContext(
    <SpecialMessage
      message={mockMessageQuestion}
      nextMessage={{
        id: 'somemessageid',
        kind: 'text',
        content: 'Answer to the question',
        created_at: '2021-01-25T20:44:00.960Z',
        conversation_id: 'fb6e60b5-ddd2-4756-a1c5-4473cc2f6a72',
        direct_message: true,
      }}
      kind="question"
      adventure={groupAdventureMock}
      step={{ ...mockStepDecisionInactive, ...{ status: 'completed' } }}
      onFocus={(): void => {
        // void
      }}
    />,
    { initialState },
  );

  expect(toJSON()).toMatchSnapshot();

  expect(queryByText(/What led you to this response today/i)).toBeTruthy();
  expect(queryByText(/Answer to the question/i)).toBeTruthy();
  expect(queryByTestId('inputEnterAnswer')).not.toBeTruthy();
  expect(queryByTestId('ctaSendAnswer')).not.toBeTruthy();
});

// TODO: FIX THIS TEST LATER
/* it('emit events correctly - kind "question"', async () => {
  const onFocus = jest.fn();
  const { store, getByTestId } = renderWithContext(
    <SpecialMessage
      message={mockMessageQuestion}
      nextMessage={null}
      kind="question"
      adventure={groupAdventureMock}
      step={mockStepDecisionInactive}
      onFocus={(): void => {
        onFocus();
      }}
    />,
    { initialState },
  );

  // Testing sending reply to the question.
  // store.clearActions();

  fireEvent(getByTestId('inputEnterAnswer'), 'onFocus');
  expect(onFocus).toHaveBeenCalled();

  fireEvent.changeText(getByTestId('inputEnterAnswer'), 'Some answer');
  fireEvent.press(getByTestId('ctaSendAnswer'));

  const actions = store.getActions();
  expect(actions).toMatchSnapshot();
});
 */
