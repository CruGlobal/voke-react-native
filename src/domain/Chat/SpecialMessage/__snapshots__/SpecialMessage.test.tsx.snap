// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly - kind "question" - answered 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "paddingBottom": 22,
      "paddingHorizontal": 22,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 20,
        },
        Object {
          "flexDirection": "column",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FF9900",
            "borderTopLeftRadius": 10,
            "borderTopRightRadius": 10,
            "paddingHorizontal": 16,
            "paddingVertical": 22,
            "width": "100%",
          },
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 17,
            "lineHeight": 22.1,
            "textAlign": "center",
          }
        }
      >
        What led you to this response today?
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderBottomLeftRadius": 10,
          "borderBottomRightRadius": 10,
          "flexDirection": "row",
          "paddingHorizontal": 16,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#1d1d26",
            },
            Object {
              "fontFamily": "TitilliumWeb-Regular",
            },
            Object {
              "color": "#186078",
              "fontSize": 17,
              "paddingVertical": 16,
              "width": "100%",
            },
          ]
        }
      >
        Answer to the question
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders correctly - kind "question" - not answered 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "paddingBottom": 22,
      "paddingHorizontal": 22,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 20,
        },
        Object {
          "flexDirection": "column",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FF9900",
            "borderTopLeftRadius": 10,
            "borderTopRightRadius": 10,
            "paddingHorizontal": 16,
            "paddingVertical": 22,
            "width": "100%",
          },
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 17,
            "lineHeight": 22.1,
            "textAlign": "center",
          }
        }
      >
        What led you to this response today?
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderBottomLeftRadius": 10,
          "borderBottomRightRadius": 10,
          "flexDirection": "row",
          "paddingHorizontal": 16,
          "width": "100%",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="sentences"
        keyboardAppearance="dark"
        multiline={true}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="enterAnswer"
        placeholderTextColor="#979797"
        rejectResponderTermination={true}
        selectionColor="#216373"
        style={
          Object {
            "color": "#186078",
            "flex": 1,
            "fontFamily": "TitilliumWeb-Regular",
            "fontSize": 17,
            "marginRight": 6,
            "paddingBottom": 22,
            "paddingTop": 22,
          }
        }
        testID="inputEnterAnswer"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "flex-end",
            "borderRadius": 88,
            "opacity": 1,
            "paddingHorizontal": 22,
            "paddingVertical": 16,
            "right": -15,
          }
        }
        testID="ctaSendAnswer"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "TitilliumWeb-Regular",
                  "textAlign": "center",
                },
                Object {
                  "fontSize": 21,
                  "lineHeight": 28.35,
                },
                Object {
                  "color": "#44c8e8",
                  "fontFamily": "TitilliumWeb-SemiBold",
                },
                undefined,
                undefined,
              ]
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": undefined,
                    "fontSize": 24,
                  },
                  Object {
                    "color": "#3295ad",
                  },
                  Object {
                    "fontFamily": "icoMoon",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              î¥€
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
